mqtt/mqtt-bridge/src/persist/memory/mod.rs

```diff
@@ -0,0 +1,218 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use async_trait::async_trait;
+use mqtt3::proto::Publication;
+use parking_lot::Mutex;
+use thiserror::Error;
+use tracing::debug;
+
+use crate::persist::{
+    memory::loader::InMemoryMessageLoader, memory::waking_map::WakingMap, Key, Persist,
+};
+
+mod loader;
+mod waking_map;
+
+/// In memory persistence implementation used for the bridge
+struct InMemoryPersist {
+    state: Arc<Mutex<WakingMap>>,
+    offset: u32,
+    loader: Arc<Mutex<InMemoryMessageLoader>>,
+}
+
+#[async_trait]
+impl<'a> Persist<'a> for InMemoryPersist {
+    type Loader = InMemoryMessageLoader;
+    type Error = InMemoryPersistError;
+
+    async fn new(batch_size: usize) -> Self {
+        let state = WakingMap::new();
+        let state = Arc::new(Mutex::new(state));
+
+        let loader = InMemoryMessageLoader::new(Arc::clone(&state), batch_size).await;
+        let loader = Arc::new(Mutex::new(loader));
+
+        let offset = 0;
+        InMemoryPersist {
+            state,
+            offset,
+            loader,
+        }
+    }
+
+    async fn push(&mut self, message: Publication) -> Result<Key, Self::Error> {
+        debug!(
+            "persisting publication on topic {} with offset {}",
+            message.topic_name, self.offset
+        );
+
+        let key = Key {
+            offset: self.offset,
+        };
+
+        let mut state_lock = self.state.lock();
+        state_lock.insert(key.clone(), message);
+        self.offset += 1;
+        Ok(key)
+    }
+
+    async fn remove(&mut self, key: Key) -> Option<Publication> {
+        debug!(
+            "removing publication with offset {} from in-flight collection",
+            self.offset
+        );
+
+        let mut state_lock = self.state.lock();
+        state_lock.remove_in_flight(&key)
+    }
+
+    async fn loader(&'a mut self) -> Arc<Mutex<InMemoryMessageLoader>> {
```

Comment:
It looks weird that it locks to iterate, but I don't know what is the solution to avoid mutex here. @dmolokanov Thoughts?